tfhe/src/core_crypto/algorithms/lwe_encryption.rs

```diff
@@ -434,6 +547,39 @@ where
     }
 }
 
+pub fn decrypt_lwe_ciphertext_non_native_mod<Scalar, KeyCont, InputCont>(
+    lwe_secret_key: &LweSecretKey<KeyCont>,
+    lwe_ciphertext: &LweCiphertext<InputCont>,
+) -> Plaintext<Scalar>
+where
+    Scalar: UnsignedInteger,
+    KeyCont: Container<Element = Scalar>,
+    InputCont: Container<Element = Scalar>,
+{
+    assert!(
+        lwe_ciphertext.lwe_size().to_lwe_dimension() == lwe_secret_key.lwe_dimension(),
+        "Mismatch between LweDimension of output ciphertext and input secret key. \
+        Got {:?} in output, and {:?} in secret key.",
+        lwe_ciphertext.lwe_size().to_lwe_dimension(),
+        lwe_secret_key.lwe_dimension()
+    );
+
+    let ciphertext_modulus = lwe_ciphertext.ciphertext_modulus();
+
+    assert!(!ciphertext_modulus.is_compatible_with_native_modulus());
+
+    let (mask, body) = lwe_ciphertext.get_mask_and_body();
+
+    Plaintext((*body.data).wrapping_sub_custom_mod(
```

Comment:
Shouldn't we do a division by the scale in the power of 2 case?
Isn't this tested?

Reply:
this is tested in the test/lwe_encryption file normally

the way the API works is that from the user perspective everything behaves 'as if' the modulus is the custom modulus, we do tricks in the background to map power of 2 cases to the native case

Reply:
this part of the code is not for power of 2 so no need to divide, though I think the naming NonNative/NonNativePowerOfTwo is maybe misleading, open to better naming